purchase_receipt_progress/models/purchase_order.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class PurchaseOrderReceivedByCategory(models.Model):
+    _name = "purchase.order.received.by.category"
+    _description = "Purchase Order Received by Category"
+
+    order_id = fields.Many2one("purchase.order")
+    category_id = fields.Many2one("product.category")
+    product_qty = fields.Float()
+    qty_received = fields.Float()
+    to_receive = fields.Float()
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    purchased_quantity_total = fields.Float(
+        string="Total Quantity", compute="_compute_purchased_quantity_total"
+    )
+    delivery_status = fields.Float(compute="_compute_delivery_status")
+    received_by_category_ids = fields.One2many(
+        "purchase.order.received.by.category",
+        "order_id",
+        string="Received by Category",
+        compute="_compute_received_by_category_ids",
+    )
+
+    @api.depends("order_line.product_qty")
+    def _compute_purchased_quantity_total(self):
+        for order in self:
+            order.purchased_quantity_total = sum(order.mapped("order_line.product_qty"))
+
+    @api.depends("order_line.product_qty", "order_line.qty_received")
+    def _compute_delivery_status(self):
+        for order in self:
+            received_quantity = sum(order.mapped("order_line.qty_received"))
+            if order.purchased_quantity_total and received_quantity:
+                order.delivery_status = (
+                    received_quantity / order.purchased_quantity_total
+                )
+            else:
+                order.delivery_status = 0
+
+    @api.depends("order_line.product_qty", "order_line.qty_received")
+    def _compute_received_by_category_ids(self):
+        for order in self:
+            purchase_order_received_by_category_obj = self.env[
+                "purchase.order.received.by.category"
+            ]
+            order_id = (
+                order._origin.id if isinstance(order.id, models.NewId) else order.id
+            )
+            purchase_order_received_by_category_obj.search(
+                [("order_id", "=", order_id)]
+            ).unlink()
+            order.received_by_category_ids = False
```

Comment:
If the idea is to remove every `purchase.order.received.by.category` before recomputing everything, it'd be better to add the `ondelete="cascade"` as I suggested above, and add `self.received_by_category_ids = False` at the beginning of the compute method. This way, the field will be emptied on every record at once (instead of emptying the field one-by-one), and the `purchase.order.received.by.category` records will all be deleted (because of the `ondelete="cascade"`).

Reply:
in any case, avoid calling search() in a loop